src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -9424,10 +9424,44 @@ public static class CExt
 ", UnmanagedCallersOnlyAttribute });
 
             comp.VerifyDiagnostics(
+                // 0.cs(7,9): error CS8901: 'SEnumerator.Dispose()' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
+                //         foreach (var i in s) {}
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "foreach (var i in s) {}").WithArguments("SEnumerator.Dispose()").WithLocation(7, 9),
                 // (14,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static non-abstract, non-virtual methods or static local functions.
                 //     [UnmanagedCallersOnly]
                 Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, "UnmanagedCallersOnly").WithLocation(14, 6)
-            );
+                );
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/73934")]
+        public void UnmanagedCallersOnlyDeclaredOnPatternDispose_CollectionExpression()
+        {
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+public struct S
+{
+    public static void M2(S s)
+    {
+        int[] a = [42, ..s];
```

Comment:
>..s [](http://example.com/codeflow?start=23&length=3)

This is an interesting scenario, but it is not quite the scenario I had in mind. I was referring to a check for a creatable collection, which involves determination of an element type (foreach element type).

Reply:
I didn't follow. How could that involve binding the `Dispose` method?

Reply:
> I didn't follow. How could that involve binding the `Dispose` method?

It looks like ```TryFindDisposePatternMethod``` might be reachable from ```TryGetCollectionIterationType```.

Reply:
Added a test. The behavior seems fine (because `TryGetCollectionIterationType` discards diagnostics)